### ISSUE TITLE
Attempt to get power device connections hangs and eventually cause out of memory and error

### ISSUE THREAD

**Issue by susan-ng** [HewlettPackard/POSH-HPOneView#28]
I’m using the “/rest/power-devices” API to automate adding maPDUs as well as configure hardware the PDU will be providing to.  Attempts to use the “/rest/power-devices” API with the powerConnections sourceConnection=0 option to get the list of hardware that can be added to the PDU hangs until the system runs out of memory and return errors.  I'm trying to get the list of hardware similar to what the "Add Connections" dialog shows from the Edit PDU dialog.  I've attached a screen capture of this hardware list in the "Add Connections dialog.

Following is the Send-HPOVRequest "/rest/power-devices" command that was run to attempt to get the list of connections available to add to the PDU, it hangs for about 25 minutes until the system runs out of memory and return the following errors and the maPDU data, but it doesn't contain the "Add Connections" hardware list:

HPONEVIEW]: Administrator@ovdcs PS C:\Users\ng> Send-HPOVRequest -uri:"/rest/power-devices" -method:GET -body:@{uuid="f1476255-faa1-4c20-9311-a9282709745d";powerConnections={connectionUri=0}}
ConvertTo-Json : Cannot convert value to type System.String.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1470 char:23
+                 $js = ConvertTo-Json -inputobject $body -Depth 99 -Compress
+                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertTo-Json], PSInvalidCastException
    + FullyQualifiedErrorId : InvalidCastFromAnyTypeToString,Microsoft.PowerShell.Commands.ConvertToJsonCommand
 
Exception calling "GetRequestStream" with "0" argument(s): "Cannot send a content-body with this verb-type."
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1475 char:11
+                 $stream = New-Object IO.StreamWriter $req.GetRequestStream()
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ProtocolViolationException
 
The property 'AutoFlush' cannot be found on this object. Verify that the property exists and can be set.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1476 char:11
+                 $stream.AutoFlush = $True
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyNotFound
 
You cannot call a method on a null-valued expression.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1477 char:11
+                 $stream.WriteLine($js)
+                 ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1478 char:11
+                 $stream.Close()
+                 ~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 

type        : ErmResourceCollection
members     : {@{deviceType=RackPdu; model=A; ratedCapacity=28; lineVoltage=4992; phaseType=SinglePhase; feedIdentifier=A; partNumber=; 
              serialNumber=; coreId=; loadSegmentId=; outletId=; managedBy=; powerConnections=System.Object[]; 
              uri=/rest/power-devices/cb318305-5161-4c19-b309-6de72a44eb2d; name=pdu-r4-1p; created=2015-07-13T04:30:23.035Z; 
              eTag=2015-07-13T04:30:23.228Z; modified=2015-07-13T04:30:23.228Z; uuid=cb318305-5161-4c19-b309-6de72a44eb2d; status=OK; 
              id=cb318305-5161-4c19-b309-6de72a44eb2d; state=Unmanaged; category=power-devices}, @{deviceType=RackPdu; model=H8B50A; 
              ratedCapacity=28; lineVoltage=208; phaseType=SinglePhase; feedIdentifier=A; partNumber=; serialNumber=; coreId=; 
              loadSegmentId=; outletId=; managedBy=; powerConnections=System.Object[]; 
              uri=/rest/power-devices/cdd41de1-c960-47e6-a53b-a6cff45ecf71; name=pdu-r1-2p; created=2015-07-15T04:18:11.362Z; 
              eTag=2015-07-15T04:18:11.412Z; modified=2015-07-15T04:18:11.412Z; uuid=cdd41de1-c960-47e6-a53b-a6cff45ecf71; status=OK; 
              id=cdd41de1-c960-47e6-a53b-a6cff45ecf71; state=Unmanaged; category=power-devices}, @{deviceType=RackPdu; model=H8B50A; 
              ratedCapacity=28; lineVoltage=208; phaseType=SinglePhase; feedIdentifier=A; partNumber=; serialNumber=; coreId=; 
              loadSegmentId=; outletId=; managedBy=; powerConnections=System.Object[]; 
              uri=/rest/power-devices/f1476255-faa1-4c20-9311-a9282709745d; name=pdu-r1-1p; created=2015-07-15T04:18:11.791Z; 
              eTag=2015-07-15T04:18:11.801Z; modified=2015-07-15T04:18:11.801Z; uuid=f1476255-faa1-4c20-9311-a9282709745d; status=OK; 
              id=f1476255-faa1-4c20-9311-a9282709745d; state=Unmanaged; category=power-devices}}
count       : 3
total       : 3
nextPageUri : 
start       : 0
prevPageUri : 
category    : power-devices
created     : 2015-07-16T01:48:28.608Z
eTag        : 
modified    : 2015-07-16T01:48:28.608Z
uri         : /rest/power-devices?start=0&count=1000 

![mapdu add connections dialog](https://cloud.githubusercontent.com/assets/11725067/8715004/a96f1382-2b49-11e5-9ced-6d389a031551.PNG)

**Comment by susan-ng** [comment]
Forgot to mention the version of OneView PowerShell Library:  "1.20.0112.0".

Thanks,
Susan Ng

**Comment by ChrisLynchHP** [comment]
Your Send-HPOVRequest call isn't correct.  PLease try this:

``` PowerShell
Send-HPOVRequest -uri "/rest/power-devices" -method GET -body @{uuid="f1476255-faa1-4c20-9311-a9282709745d";powerConnections=@{connectionUri=0}}
```

You are missing an @ sign in your powerConnections property value.

**Comment by susan-ng** [comment]
Hi Chris,

Thank you very much for your prompt response and help!

I copy and pasted your Send-HPOVRequest command and it no longer hangs or use up the system memory, but it still returns errors along with the data.  The powerConnections property doesn't contain the list of resources that can be configured to the maPDU:

[HPONEVIEW]: Administrator@ovdcs.cloud.internal PS C:\Users\ng> $c=Send-HPOVRequest -uri "/rest/power-devices" -method GET -body @{uuid="f1476255-faa1-4c20-9311-a9282709745d";powerConnections=@{connectionUri=0}}
Exception calling "GetRequestStream" with "0" argument(s): "Cannot send a content-body with this verb-type."
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1475 char:11
+                 $stream = New-Object IO.StreamWriter $req.GetRequestStream()
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ProtocolViolationException
 
The property 'AutoFlush' cannot be found on this object. Verify that the property exists and can be set.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1476 char:11
+                 $stream.AutoFlush = $True
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyNotFound
 
You cannot call a method on a null-valued expression.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1477 char:11
+                 $stream.WriteLine($js)
+                 ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At C:\HPCS\CI\Common\HPOneView.120\HPOneView.120.psm1:1478 char:11
+                 $stream.Close()
+                 ~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
[HPONEVIEW]: Administrator@ovdcs.cloud.internal PS C:\Users\ng> $c.members


deviceType       : RackPdu
model            : A
ratedCapacity    : 28
lineVoltage      : 4992
phaseType        : SinglePhase
feedIdentifier   : A
partNumber       : 
serialNumber     : 
coreId           : 
loadSegmentId    : 
outletId         : 
managedBy        : 
powerConnections : {}
uri              : /rest/power-devices/cb318305-5161-4c19-b309-6de72a44eb2d
name             : pdu-r4-1p
created          : 2015-07-13T04:30:23.035Z
eTag             : 2015-07-13T04:30:23.228Z
modified         : 2015-07-13T04:30:23.228Z
uuid             : cb318305-5161-4c19-b309-6de72a44eb2d
status           : OK
id               : cb318305-5161-4c19-b309-6de72a44eb2d
state            : Unmanaged
category         : power-devices

deviceType       : RackPdu
model            : H8B50A
ratedCapacity    : 28
lineVoltage      : 208
phaseType        : SinglePhase
feedIdentifier   : A
partNumber       : 
serialNumber     : 
coreId           : 
loadSegmentId    : 
outletId         : 
managedBy        : 
powerConnections : {}
uri              : /rest/power-devices/cdd41de1-c960-47e6-a53b-a6cff45ecf71
name             : pdu-r1-2p
created          : 2015-07-15T04:18:11.362Z
eTag             : 2015-07-15T04:18:11.412Z
modified         : 2015-07-15T04:18:11.412Z
uuid             : cdd41de1-c960-47e6-a53b-a6cff45ecf71
status           : OK
id               : cdd41de1-c960-47e6-a53b-a6cff45ecf71
state            : Unmanaged
category         : power-devices

deviceType       : RackPdu
model            : H8B50A
ratedCapacity    : 28
lineVoltage      : 208
phaseType        : SinglePhase
feedIdentifier   : A
partNumber       : 
serialNumber     : 
coreId           : 
loadSegmentId    : 
outletId         : 
managedBy        : 
powerConnections : {}
uri              : /rest/power-devices/f1476255-faa1-4c20-9311-a9282709745d
name             : pdu-r1-1p
created          : 2015-07-15T04:18:11.791Z
eTag             : 2015-07-15T04:18:11.801Z
modified         : 2015-07-15T04:18:11.801Z
uuid             : f1476255-faa1-4c20-9311-a9282709745d
status           : OK
id               : f1476255-faa1-4c20-9311-a9282709745d
state            : Unmanaged
category         : power-devices

Thanks,
Susan

**Comment by ChrisLynchHP** [comment]
I didn't notice you were doing a GET HTTP method.  You cannot provide a body when you are issuing a GET method.  You can only with POST, PUT or PATCH HTTP methods.  If you are trying to get a PDU resource object, your command needs to be:

``` powershell
Send-HPOVRequest -Uri "/rest/power-devices/f1476255-faa1-4c20-9311-a9282709745d" 
```

You don't need to include the semi-colon (:) when specifying values for parameters.  You only do that when the parameter is either Boolean or Switch type, and then you provide either $True or $False.

**Comment by susan-ng** [comment]
Hi Chris,

Since the GET HTTP method cannot provide a body, what is the correct syntax to get a list of hardware resources that can be configured to the maPDU?  

Based on the OneView API Reference Guide, I can use the query parameter "fields=powerConnections" on the GET HTTP method but it just returns the same full set of data as without this parameter.  I'm not sure how to specify the "powerConnections" sub-option "sourceConnection=0" in the GET HTTP method since this seems to be the option to show the list of discovered connections.  Following is the description from the API Reference Guide:


powerConnections     The set of resources powered by this power delivery device.  
 
     connectionUri            The uri of the resource that is connected to the power delivery device.  
 
     deviceConnection     The power supply number of the connection to the resource.  
 
     sourceConnection     An optional connection on the power delivery device. A value of 0 indicates a discovered connection. Manually specified source connections should be greater than zero.   
 
Following is the return data from using the "fields=powerConnections" parameter in the GET HTTP request:

[HPONEVIEW]: Administrator@ovdcs.cloud.internal PS C:\Users\ng> Send-HPOVRequest -Uri "/rest/power-devices/f1476255-faa1-4c20-9311-a9282709745d?fields=powerConnections"


deviceType       : RackPdu
model            : H8B50A
ratedCapacity    : 28
lineVoltage      : 208
phaseType        : SinglePhase
feedIdentifier   : A
partNumber       : 
serialNumber     : 
coreId           : 
loadSegmentId    : 
outletId         : 
managedBy        : 
powerConnections : {}
uri              : /rest/power-devices/f1476255-faa1-4c20-9311-a9282709745d
name             : pdu-r1-1p
created          : 2015-07-15T04:18:11.791Z
eTag             : 2015-07-15T04:18:11.801Z
modified         : 2015-07-15T04:18:11.801Z
uuid             : f1476255-faa1-4c20-9311-a9282709745d
status           : OK
id               : f1476255-faa1-4c20-9311-a9282709745d
state            : Unmanaged
category         : power-devices

Thank you for your help!

Regards,
Susan

**Comment by ChrisLynchHP** [comment]
This is fixed in [Release 1.20.208.0](https://github.com/HewlettPackard/POSH-HPOneView/releases/tag/v1.20.208.0)
